File: form-runner/jvm/src/main/assets/xbl/orbeon/countdown/countdown.js
/**
 * Copyright (C) 2017 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
(function() {

    var $ = ORBEON.jQuery;
    var _ = ORBEON._;

    YAHOO.namespace('xbl.fr');
    YAHOO.xbl.fr.Countdown = function() {};
    ORBEON.xforms.XBL.declareClass(YAHOO.xbl.fr.Countdown, 'xbl-fr-countdown');

    var lastTimestamp = Date.now();
    var outputEls = [];

    function parseDuration(text) {
        var minSec = text.split(":");
        if (minSec.length == 1) {
            return parseInt(minSec[0]);
        } else if (minSec.length == 2) {
            var min = parseInt(minSec[0]);
            var sec = parseInt(minSec[1]);
            if (_.isNaN(min) || _.isNaN(sec)) return NaN;
            return min*60 + sec;
        } else {
            return NaN;
        }

    }

    function serializeDuration(secTotal) {
        var min = Math.floor(secTotal / 60);
        var sec = secTotal % 60;
        var secPart = sec < 10 ? "0" + sec : sec;
        return min + ":" + secPart;
    }

    setInterval(function() {
        var newTimestamp = Date.now();
        var increment = Math.floor((newTimestamp - lastTimestamp)/1000);
        if (increment > 0) {
            lastTimestamp = newTimestamp;
            _.each(outputEls, function(outputEl) {
                var sec = parseDuration($(outputEl).text());
                if (! _.isNaN(sec)) {
                    var countdown = YAHOO.xbl.fr.Countdown.instance(outputEl);
                    var containerId = countdown.container.id;
                    var newSec = sec - increment;
                    if (newSec <= countdown.alertThreshold && ! $(outputEl).is(".fr-countdown-alert")) {
                        $(outputEl).addClass("fr-countdown-alert")
                        ORBEON.xforms.Document.dispatchEvent(containerId, "fr-countdown-alert");
                    }
                    if (newSec >= 0) {
                        var newDuration = serializeDuration(newSec);
                        $(outputEl).text(newDuration);
                    }
                    if (newSec == 0) {
                        ORBEON.xforms.Document.dispatchEvent(containerId, "fr-countdown-ended");
                    }
                }
            });
        }
    }, 100);

    YAHOO.xbl.fr.Countdown.prototype = {

        outputEl: null,
        alertThreshold: null,

        init: function() {
            this.outputEl = $(this.container).find(".xforms-output-output").get(0);
            if (! _.contains(outputEls, this.outputEl))
                outputEls.push(this.outputEl);
        },

        setAlertThreshold: function(alertThreshold) {
            this.alertThreshold = parseInt(alertThreshold) * 60;
        },

        durationChanged: function(newValue) {
            $(this.outputEl).text(newValue);
            var sec = parseDuration(newValue);
            if (sec > this.alertThreshold)
                $(this.outputEl).removeClass("fr-countdown-alert")
        }
    };
})();

File: form-runner-web/src/main/scala/org/orbeon/xbl/Countdown.scala
package org.orbeon.xbl

import org.orbeon.web.DomSupport.*
import org.orbeon.xforms.facade.XBLCompanion
import org.orbeon.xforms.{AjaxClient, AjaxEvent, XFormsXbl}
import org.scalajs.dom.html

import scala.collection.mutable
import scala.scalajs.js


object Countdown {

  XFormsXbl.declareCompanion("fr|countdown", js.constructorOf[CountdownCompanion])

  private val companions = mutable.Set[CountdownCompanion]()

  private var lastTimestamp = js.Date.now()

  js.timers.setInterval(100) {
    val newTimestamp = js.Date.now()
    val increment = Math.floor((newTimestamp - lastTimestamp) / 1000).toInt
    if (increment > 0) {
      lastTimestamp = newTimestamp
      // Create a copy to avoid ConcurrentModificationException if destroy is called during iteration
      companions.toList.foreach(_.tick(increment))
    }
  }

  private def parseDuration(text: String): Option[Int] = {
    val minSec = text.split(":")
    if (minSec.length == 1) {
      minSec(0).toIntOption
    } else if (minSec.length == 2) {
      for {
        min <- minSec(0).toIntOption
        sec <- minSec(1).toIntOption
      } yield min * 60 + sec
    } else {
      None
    }
  }

  private def serializeDuration(secTotal: Int): String = {
    val min = Math.floor(secTotal / 60).toInt
    val sec = secTotal % 60
    val secPart = if (sec < 10) s"0$sec" else s"$sec"
    s"$min:$secPart"
  }

  private class CountdownCompanion(containerElem: html.Element) extends XBLCompanion {

    private var outputElOpt: Option[html.Element] = None
    private var alertThreshold: Int = 0

    override def init(): Unit = {
      outputElOpt = containerElem.querySelectorOpt(".xforms-output-output")
      companions += this
    }

    override def destroy(): Unit =
      companions -= this

    def setAlertThreshold(threshold: String): Unit =
      alertThreshold = threshold.toIntOption.getOrElse(0) * 60

    def durationChanged(newValue: String): Unit = {
      outputElOpt.foreach(_.innerText = newValue)
      val sec = parseDuration(newValue).getOrElse(0)
      if (sec > alertThreshold)
        outputElOpt.foreach(_.classList.remove("fr-countdown-alert"))
    }

    def tick(increment: Int): Unit =
      for {
        outputEl <- outputElOpt
        sec      <- parseDuration(outputEl.innerText)
      } locally {
        val newSec = sec - increment
        if (newSec <= alertThreshold && !outputEl.classList.contains("fr-countdown-alert")) {
          outputEl.classList.add("fr-countdown-alert")
          AjaxClient.fireEvent(AjaxEvent("fr-countdown-alert", containerElem.id))
        }
        if (newSec >= 0)
          outputEl.innerText = serializeDuration(newSec)
        if (newSec == 0) {
          AjaxClient.fireEvent(AjaxEvent("fr-countdown-ended", containerElem.id))
        }
      }
  }
}
